llvm/lib/CodeGen/SelectionDAG/LegalizeDAG.cpp

```diff
@@ -1524,11 +1524,14 @@ SDValue SelectionDAGLegalize::ExpandConcatVectors(SDNode *Node) {
   SmallVector<SDValue, 16> Ops;
   unsigned NumOperands = Node->getNumOperands();
   MVT VectorIdxType = TLI.getVectorIdxTy(DAG.getDataLayout());
+  EVT VectorValueType = Node->getOperand(0).getValueType();
+  EVT ElementValueType = VectorValueType.getVectorElementType();
+  if (ElementValueType !=
+      TLI.getTypeToTransformTo(*DAG.getContext(), ElementValueType)) {
```

Comment:
I think checking whether the extract_vector_elt will be Expand would make more sense than checking the type's transformed type

Reply:
This is similar to what justinfargnoli suggested, but I believe the extract_vector_elt operation is legal for the particular test case that is worse. Even when I made the check:

```  
  if (!TLI.isOperationLegal(ISD::EXTRACT_VECTOR_ELT, ElementValueType)) {
      return ExpandVectorBuildThroughStack(Node);
  }
```

I didn't see a change, so it seems like just checking the action associated with the operation is not sufficient.